### DISCUSSION TITLE
Git-Sync Containers Keep failing and restarting

### DISCUSSION THREAD

**Discussion by ahipp13** [apache/airflow#30469] [Q&A] [answered]
Hi all,

Am writing this discussion to ask some help from the Airflow community. Currently we are having some weird behavior with the git sync container on the scheduler and triggerer pods and want to get some insights from people to see if they think this is an Airflow issue or something else.

To start, we are running Airflow 2.5.1, and are using Kubernetes and the Official Airflow Helm Chart 1.8.0, with an external postgres backend database. We are also using a git sync sidecar to store our DAGs, and to get Airflow to access this repository, we updated the _helpers.yaml file and values file to put in the username and password.
![image](https://user-images.githubusercontent.com/118911990/229902171-6d6f8ef9-09fe-425c-83d9-07987c9e5b39.png)
We currently have 3 environments running this Airflow setup (prod, test, dev). All the environments are basically the same except for they each do have their own DAG repository and have different DAGs in them. 

What is happening is at the exact same time, everyday, for some reason the git-sync containers for the scheduler and triggerer go crazy and just start restarting out of no where:
![image](https://user-images.githubusercontent.com/118911990/229903439-7ac832ed-1cd2-4f5b-b78d-423a4d983a52.png)
And this exact same thing happens in all of our environments. It has these episodes at approximently 5:40 and 7:40 UTC time, and then the rest of the day it is perfectly fine.

Looking at the logs from the git-sync containers this is what it shows: 
![image](https://user-images.githubusercontent.com/118911990/229906560-800a36ec-0750-4dd1-90c1-a602c025cfae.png)
it shows a 429, which is the error code for too many requests. 

At first I thought it was maybe the DAGs, so I ran all of them in each env at once to see if that would cause it. Nothing happened, and there were no restarts.

Then what I did was changed the wait time in our dev env from 5 seconds to 15 seconds and then checked it the next morning. It reduced the restarts, but 2 still did happen at that same time frame, 7:40 UTC:
![image](https://user-images.githubusercontent.com/118911990/229905085-db375f18-b8c1-4bd0-8225-74a4968e010a.png)

So my question is, is this something wrong with the Airflow git sync, or is something that we do not know about accessing our repos at this time?
Really, I am just trying to determine if this is an Airflow issue, or if this is something outside of Airflow.

@potiuk or anybody else I would really appreciate some insight on this. If you have any questions or need more information please feel free to ask. Thank you!

**Comment by potiuk** [comment]
Error 429 is "too many requests" - https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429 - you seem to exceed the limits.

The "timing" of the issue indicate that there is something that uses the same git credentials around that time that overwhelms GIT with requests. It does not seem it comes from Airflow - because you would have this problem all the time, but very likely around that time you are running some massive GitHub sync or smth that exceed the limits. 

Unless the problem is that you are starting many tasks around the same time and they are generating massive number of requests when they start quickly around the same time.

https://docs.github.com/en/rest/overview/resources-in-the-rest-api?apiVersion=2022-11-28#rate-limiting

Rate limits for GitHub are applied "per authenticated user" when the user is authenticated, so the solution to that is to have a separate user only used by Airflow and NOT to share the credentials with some other systems of yours (especially if they generate a lot of HTTP requests".

**Reply by ahipp13** [reply]
Wanted to test this theory out so I switched our dev environment to use my PAT token (which is not involved in any other processes) and still got restarts at the same times:
![image](https://user-images.githubusercontent.com/118911990/230084911-d60e3c99-db9c-4309-ba51-1fc4d8f342d2.png)

Ran a command to check the rate limits on my token (it is almost the end of an hour so it should so some uses) and this is the result
 curl -L -H "Accept: application/vnd.github+json" -H "Authorization: Bearer MY_TOKEN" -H "X-GitHub-Version: 2022-11-28"  https://api.github.com/rate_limit
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1193  100  1193    0     0   2348      0 --:--:-- --:--:-- --:--:--  2362{
  "resources": {
    "core": {
      "limit": 5000,
      "used": 0,
      "remaining": 5000,
      "reset": 1680702743
    },
    "search": {
      "limit": 30,
      "used": 0,
      "remaining": 30,
      "reset": 1680699203
    },
    "graphql": {
      "limit": 5000,
      "used": 0,
      "remaining": 5000,
      "reset": 1680702743
    },
    "integration_manifest": {
      "limit": 5000,
      "used": 0,
      "remaining": 5000,
      "reset": 1680702743
    },
    "source_import": {
      "limit": 100,
      "used": 0,
      "remaining": 100,
      "reset": 1680699203
    },
    "code_scanning_upload": {
      "limit": 1000,
      "used": 0,
      "remaining": 1000,
      "reset": 1680702743
    },
    "actions_runner_registration": {
      "limit": 10000,
      "used": 0,
      "remaining": 10000,
      "reset": 1680702743
    },
    "scim": {
      "limit": 15000,
      "used": 0,
      "remaining": 15000,
      "reset": 1680702743
    },
    "dependency_snapshots": {
      "limit": 100,
      "used": 0,
      "remaining": 100,
      "reset": 1680699203
    }
  },
  "rate": {
    "limit": 5000,
    "used": 0,
    "remaining": 5000,
    "reset": 1680702743
  }
}

So I am pretty stumped.... Maybe there is some process we are unaware of that accesses our repos at this time and it is causing us not to be able to send requests? Because it seems like it is not a rate issue since I have tried 2 different PATs now and they both lead to the same result. We only have one DAG running at the time of the restarts, and yesterday we ran that DAG during the day to see if it was causing it, and there were no restarts... So it did not seem to be that DAG.

**Reply by potiuk** [reply]
Are you on enterprise network? If so - maybe you have a proxy (transparent) in-between you and github and that's the one returning 429? Then you should likely talk to your admins. 

If not - i suggest to open an issue to GitHub. You are on enterprise support level I think (guessing seeing 15000 so they shoudl respond pretty quickly - they usually do).

**Reply by ahipp13** [reply]
I am, and so that is the only thing I can think of that would be the cause of it. We increased the wait time from 5s to 15s, and the amount of restarts went down, but did not go away completely:

![image](https://user-images.githubusercontent.com/118911990/230616571-101385c5-0dd7-4371-99fe-496699abaac5.png)

So the only thing I can really think of at this point is there must be a ton of traffic from our network at that specific time. We are going to update our git-sync image to the latest one to see if that gets rid of the remaining restarts. If it doesn't we will try and increase the wait time again to see if it gets rid of them. I will provide another update after these have been done. Thank you for your help in discussing this with me!

**Reply by potiuk** [reply]
OK. Let me know how it plays out.

**Reply by ahipp13** [reply]
To get back on this, you were correct: it is our enterprise network, our proxy is sending too many unauthorized requests which makes the 429s. Interestingly, it is saying that our team as well as others is making unauthorized requests using this URL: “GITHUB_DAG_REPO.git/info/refs?service=git-upload-pack"

Now as you know the git sync runs this command: "git ls-remote -q GITHUB_DAG_REPO.git refs/heads/main" and in order for us to even get it to work we had to add a user and a PAT token. So why git-upload-pack is making unauthorized requests, I have no idea... I have responded to the GitHub support guy that made us aware of this, I will provide an update when he answers

**Reply by potiuk** [reply]
Very interesting indeed . I might also take a look knowing all the details. Maybe i can find something

**Reply by potiuk** [reply]
I think this is the same issue as here: https://github.com/kubernetes/git-sync/issues/522 -> though it has been closed due to lack of reproducibility, but it seems rather similar.

**Reply by ahipp13** [reply, marked as answer]
Interesting read. After talking more with the GitHub vendor, it seems as if what I said before may not be the issue, and after the limit is hit we may have just been lumped in with everybody else. Which is why it looked like we were making unauthed requests even though we are not. At the EOD, the reason it is happening is because every team is lumped together on one IP, and there are some teams that are making tons of unauthed requests. I lowered our git sync wait time to 30s and now we only see 1 or 2 restarts if that, so that is what we are going with for now. I have been engaged with the our GitHub team so I am hoping they will find whoever is making all the requests or create multiple IPs so that this does not happen. But I am going to mark this as the answer since I myself cannot do anything else!

**Reply by potiuk** [reply]
Thanks, yeah. That might well be.
